Change focus management in disclosure utility component

alex-ju opened the pull request:
### :pushpin: Summary

This PR is proposing an alternative approach to focus management in our disclosure utility component on the back of #280

### :hammer_and_wrench: Detailed description

<!-- If more details are appropriate, add them here. What code changed, and why? -->
The main change is dropping the focus-trap mechanism and letting the focus flow through the component while preserving the rest of the requirements defined in [the associated CRD](https://docs.google.com/document/d/1PSIAdUypvGWCLBAkeOo4HSa3qd6qN-pWvlLrfuwYT4o/edit#). The solution is meant to avoid any change the API of the component.

I propose this change to fix the issues described in #280 (briefly: "fix focus-trap bug when the interactive opens a modal", and the "scrolling caused by focus returning to the previously active element"). It also addresses the WCAG 2.1 success criterion [2.1.2 No Keyboard Trap](https://www.w3.org/TR/WCAG21/#no-keyboard-trap) (Level A) by allowing the users to exit the disclosed content using the default keyboard shortcut (tab).

### :camera_flash: Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td>

![disclosure-focus-before](https://user-images.githubusercontent.com/788096/168081360-26c78968-aa97-4c52-b6d7-97bc07b529ec.gif)

</td><td>

![disclosure-focus-after](https://user-images.githubusercontent.com/788096/168081408-cd8ab2ae-80a1-4760-bb0b-0dc5c7dc620b.gif)

</td></tr></table>

Note: the colors went way off on the above compressed gifs 😱 

### :link: External links

[Review of focus management in Disclosure (+ Dropdown/Breadcrumb) components](https://github.com/hashicorp/design-system/issues/280)
Fixes https://github.com/hashicorp/design-system/issues/267 and https://github.com/hashicorp/cloud-ui/pull/2362#issuecomment-1097778784

***

### 👀 How to review

<!-- Suggest how it's best for the reviewer to review the code (choose one, or remove) -->
👉 Review commit-by-commit

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202253786016960

MelSumner commented:
Thank you for your PR! Can you please add some more details to the PR, explaining why you want to make this change, and what the intent is. Thank you!

didoo reviewed (changes requested):
@alex-ju I've found an issue with Safari that I suspect is a blocker: when the dropdown is activated/opened via keyboard (tab to the toggle, open the toggle via spacebar/enter) everything works as expected; but if the dropdown is opened via mouse click, then nor the click outside nor the "esc" key close the dropdown.

other things I notice, that may be helpful in narrowing down the problem: 
- if you add a `console.log` inside the `onFocusOut` function, you will see that is logged in Chrome (when the toggle is clicked to open it) but not in Safari
- in Safari, if you open the toggle via mouse click, but then hit `tab` and focus on the first item, then the click outside and the esc both trigger the close 

I did some tests, but I am not sure what would be the cause root, if something in the logic or a bug in Safari. I'll let you investigate.

This happens in Safari:
<img width="887" alt="screenshot_1448" src="https://user-images.githubusercontent.com/686239/168310285-ecbfbe30-3f44-49d2-87fe-ad62a8eba1ca.png">

alex-ju commented:
> Thank you for your PR! Can you please add some more details to the PR, explaining why you want to make this change, and what the intent is. Thank you!

Sure, I should've been more clear on the why; I guess part of it was left in [the comment on the issue](https://github.com/hashicorp/design-system/issues/280#issuecomment-1124978813). I have added a paragraph to the PR description trying to articulate the motivation behind the proposed change.

alex-ju commented:
> @alex-ju I've found an issue with Safari that I suspect is a blocker: when the dropdown is activated/opened via keyboard (tab to the toggle, open the toggle via spacebar/enter) everything works as expected; but if the dropdown is opened via mouse click, then nor the click outside nor the "esc" key close the dropdown.
> 
> other things I notice, that may be helpful in narrowing down the problem:
> 
> * if you add a `console.log` inside the `onFocusOut` function, you will see that is logged in Chrome (when the toggle is clicked to open it) but not in Safari
> * in Safari, if you open the toggle via mouse click, but then hit `tab` and focus on the first item, then the click outside and the esc both trigger the close
> 
> I did some tests, but I am not sure what would be the cause root, if something in the logic or a bug in Safari. I'll let you investigate.

It's [a bug in webkit](https://trac.webkit.org/changeset/276698/webkit/), [a fix will be released soon](https://webkit.org/blog/11680/release-notes-for-safari-technology-preview-125/), but I'll add a patch in the meantime.

alex-ju commented:
@didoo I've added a patch for the inconsistency of WebKit not triggering `focusin` and `focusout` on button elements (among other types of elements) to address the issue you found and added your suggestions to the preview/documentation page.

didoo commented on packages/components/addon/components/hds/disclosure/index.js:
```diff
@@ -7,41 +7,47 @@ export default class HdsDisclosureComponent extends Component {
       this.toggleRef = event.currentTarget;
     }
     this.isActive = !this.isActive;
+    this.toggleRef.focus();
```
[suggestion] generally, I tend to add comments to this kind of "things" (when I make a non-obvious fix for two reasons: a future me that will come in and don't remember why I did something + the fear that a future maintainer will come in and remove the code because they think is unnecessary).

didoo reviewed (approved):
All good for me 👍

MelSumner commented:
Is this a proposal or a foregone conclusion? I think I missed that part of the conversation entirely. Any documents that you can point me to? I don't think we should change the component without appropriate justification.

MelSumner reviewed (approved):
Update: After looking through it, I really think we are losing some A11y-specific UX here, but it's also technically a WCAG-conformant solution. As such, I disagree with this change but cannot object to it.

alex-ju commented:
> Update: After looking through it, I really think we are losing some A11y-specific UX here, but it's also technically a WCAG-conformant solution. As such, I disagree with this change but cannot object to it.

Thank you for the review, Melanie. From an accessibility perspective, my concern with the current implementation is that assistive tech users are not being told that they've entered a region when they moved into the disclosed content and they're not being told that if they want to get out of the region they can use the escape key (some may figure it out, but it's not explicit), which may trap and frustrate some of them. That being said, we'll most probably need this focus-trap-based solution for the modal dialogue component.
